### ISSUE TITLE
Debugging overhaul and redesign

### ISSUE THREAD

**Issue by filipdutescu** [helix-editor/helix#5950]
# Description

The current debugging experience is less than ideal. While challenges are expected with TUI editors, the current experience is, to put it bluntly, raw. It is hard to launch or stop sessions, view variables, see which line the session is paused at etc. You also have no way to drill down into variable values (or if you have, it is so hidden I have yet to find it) or edit their values during a debugging session.

There is much more that should be done to ensure a smooth experience, including:
1. defaults for binary targets (ie configuration type - Debug/Release, binary location - `target/debug/myexe`) which should be overridable in `language.toml` as well
2. attach should show a searchable list of processes the user could choose from (search should include the option for process ID)
3. #5927

# Proposal

To address some of those pain points, I propose the following overhaul and redesign:
- [ ] Add a new editor mode, alongside `NOR`, `INS`, `SEL`: the `DBG` mode, which would be basically `NOR` with extra configurable keybindings for the debugging commands. This would also remove the need for the sticky popup, which could be turned on by a new keybinding, such as `?`, for example;
- [ ] Highlight the current line at which execution is paused, This also means adding new theme colors for breakpoints and the current line;
- [ ] Add a new debug panel, which should be able to be placed in all cardinal points (left, top, bottom, right), containing at least: a variables panel (locals, arguments, registers etc.), watches, call stack and breakpoints. Those individual modules should also be able to be placed in any of the 4 directions by themselves, in order to have a more personal and customized UI. The variables should be expandable, showing their inner fields (private, protected and public) with their values. Those "children" should behave the same, if they are complex types. Watches and variable values should be editable with a prompt.
- [ ] A quick watch popup should be made, which could start empty or from the current code selection, behaving similar to a picker, with expandable values for the result of the expression.

# Related issues

* #5927

# UI/UX mockups

New UI for debugging:
![image](https://user-images.githubusercontent.com/46979062/218330881-411619b4-fe52-4088-8df5-5d681ffbdfc2.png)

Help popup (using the current one as example):
![image](https://user-images.githubusercontent.com/46979062/218330925-d1a22744-e5f8-4f25-8aee-378d2cf3b016.png)

Expanding a variable's value (same prompt and cursor highlighting should be used for watches as well):
![image](https://user-images.githubusercontent.com/46979062/218330961-0a5a39e9-d494-49d6-8060-e583057ccf0d.png)

Quick watch popup:
![image](https://user-images.githubusercontent.com/46979062/218331018-fa833b12-9b97-42eb-8f39-3a5fa5dcafdb.png)

**Comment by archseer** [comment]
For the variable views (the top pane called Locals -- it's technically variables), the idea was to repurpose the tree component from https://github.com/helix-editor/helix/pull/2377/https://github.com/helix-editor/helix/pull/5768

**Comment by filipdutescu** [comment]
> For the variable views (the top pane called Locals -- it's technically variables), the idea was to repurpose the tree component from https://github.com/helix-editor/helix/pull/2377 / https://github.com/helix-editor/helix/pull/5768

Yes, I plan on using the exact same mechanism as proposed by you. That part (local pane) and the quick watch one would be dependent on either of the two linked PRs.

**Comment by filipdutescu** [comment]
@CptPotato proposed a new design for the debug panel,which I like quite a bit. This is what I propose, moving forward:
![image](https://user-images.githubusercontent.com/46979062/218580023-b26888ab-b65f-433a-abc7-fa968bde2744.png)

**Comment by filipdutescu** [comment]
Update on the current debug line (#5952). Latest developments lead me to this current implementation, which is a good MVP in my opinion. Further work can be done to improve the ability to customise components such as selection and cursor highlights, but they'll be out of scope for the current PR.

![image](https://user-images.githubusercontent.com/46979062/221024396-bd94ddf4-d3aa-4c6d-abfa-5a46c261a787.png)

**Comment by filipdutescu** [comment]
Update on the default theme. After talking some more with @archseer, we landed on this colour for the debug highlight, which I find beautiful:

![image](https://user-images.githubusercontent.com/46979062/224116478-0d041623-f577-4156-a4e9-e30fe70966d3.png)

**Comment by filipdutescu** [comment]
Latest update: #5952 was merged, hope it is a pleasant addition and a nice start to the overhaul! Thanks a lot to @pascalkuthe and @archseer for the help with the issue!

**Comment by teburd** [comment]
Would a disasm split or register view be something to consider? Debugging low level code these features in gdb are incredibly helpful

**Comment by andreciobezerra** [comment]
I hope this proposal is going on, the helix is fantastic, but the debug is really experimental.

**Comment by filipdutescu** [comment]
I have had some personal problems keeping me from dedicating sufficient time to tuis initiative. But all should start to clear up during summer, when I plan to get up to date with any progress made and pick it back up. Not sure what progress was made in the meantime though.

**Comment by gyreas** [comment]
Adding mouse actions like resizing the left pane would be nice, making the side pane windows collapsible would be another awesome addition too. By the way, who's had any luck using the debugger?

**Comment by Rot127** [comment]
To throw another "good to have" feature on the heap, it would be great if the interface supports data viewers of different languages.

E.g. having `C` + `Rust` is pretty common. And there are quite some bugs which occur during the interaction of the two.
If the debugger switches from a Rust file to a C file, it should be able to seamlessly show the C types.

I just can't find an IDE which does this without problems (RustRover doesn't show C types and in VSCode, Rust debugging is broken with complicated building scenarios).

**Comment by taylorpool** [comment]
Curious what the progress has been here. I love Helix and debugging!

**Comment by johan-lejdung** [comment]
This addition would be greatly appreciated! The current implementation is just not enough, I cannot list the values for the most part as they overflow from the tiny box I'm presented with. 

Is there any active work on implementing this or is it still in the idea stage?

**Comment by IanQS** [comment]
For python there's an issue tracking the current status and what people are observing trying to use it https://github.com/helix-editor/helix/discussions/12722

**Comment by git-staus** [comment]
We users would love to try out the debugger, but the need for documentation as described in #4231 is still as pressing as it was three years ago. If somebody could perhaps write some instructions on how to use the functionality as it currently exists, it might pull in more users / contributors / eyeballs to help move this forward.

**Comment by IanQS** [comment]
@git-staus check out: https://github.com/helix-editor/helix/discussions/12722#discussioncomment-17172919

**Comment by jasonwilliams** [comment]
I know this thread is old but would love to see some love go into the debugger. It's very unsuable at the moment, there seems to be many places where it just locks up completely.

My focus has mainly been on enabling JavaScript debugging, but the [fixes](https://github.com/helix-editor/helix/pull/13403) I've provided help all DAP adapters. 

I like the mockups here and the proposal, hopefully we can have more people help in this area
